internal/async/mq_client.go

```diff
@@ -0,0 +1,89 @@
+package async
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+
+	"github.com/nats-io/nats.go"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	SubmitTxTopic   = "submit-tx"
+	MinedTxsTopic   = "mined-txs"
+	RegisterTxTopic = "register-tx"
+	RequestTxTopic  = "request-tx"
+)
+
+type NatsClient interface {
+	QueueSubscribe(subj, queue string, cb nats.MsgHandler) (*nats.Subscription, error)
+	Close()
+	Publish(subj string, data []byte) error
+	Drain() error
+}
+
+type MQClient struct {
+	nc     NatsClient
+	logger *slog.Logger
+}
+
+func WithLogger(logger *slog.Logger) func(handler *MQClient) {
+	return func(m *MQClient) {
+		m.logger = logger
+	}
+}
+
+func NewNatsMQClient(nc NatsClient, opts ...func(client *MQClient)) *MQClient {
+	m := &MQClient{
+		nc:     nc,
+		logger: slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{Level: slog.LevelInfo})),
+	}
+	for _, opt := range opts {
+		opt(m)
+	}
+
+	return m
+}
+
+func (c MQClient) Shutdown() {
+	if c.nc != nil {
+		err := c.nc.Drain()
+		if err != nil {
+			c.logger.Error("failed to drain nats connection", slog.String("err", err.Error()))
+		}
+	}
+}
+
+func (c MQClient) Publish(topic string, hash []byte) error {
```

Comment:
Maybe `PublishHash` would be a better name? `Publish` suggests like you could publish any message. WDYT?

Reply:
Ok, then I'd rather change the name of the byte array. Because I the message queue client should be agnostic on what it publishes. It can be a hash or something else. Agree?

```
 Publish(topic string, data []byte)
```

Reply:
Okay, agree. I thought it's specific to publishing only hashes